src/core/server/saved_objects/migrationsv2/model.ts

```diff
@@ -626,59 +702,122 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
           lastHitSortValue: res.right.lastHitSortValue,
         };
       } else {
+        // we don't have any more outdated documents and need to either fail or move on to updating the target mappings.
+        if (stateP.corruptDocumentIds.length > 0 || stateP.transformErrors.length > 0) {
+          // If documents couldn't be transformed or there were transformation errors we fail the migration
+          const corruptDocumentIdReason =
+            stateP.corruptDocumentIds.length > 0
+              ? `The following corrupt saved object documents: ${stateP.corruptDocumentIds.join(
+                  ','
+                )}`
+              : '';
+          // we have both the saved object Id and the stack trace in each `transformErrors` item.
+          const transformErrorsReason =
+            stateP.transformErrors.length > 0
+              ? 'the following saved object documents could not be transformed:/n' +
+                stateP.transformErrors
+                  .map((errObj) => `${errObj.rawId}: ${errObj.err.message}`)
+                  .join('/n')
+              : '';
+          return {
+            ...stateP,
+            controlState: 'FATAL',
+            reason: `Migrations failed. Reason: ${corruptDocumentIdReason} ${transformErrorsReason}. To allow migrations to proceed, please delete these documents.`,
+          };
+        } else {
+          // If there are no more results we have transformed all outdated
+          // documents and we didn't encounter any corrupt documents or transformation errors
+          // and can proceed to the next step
+          return {
+            ...stateP,
+            controlState: 'OUTDATED_DOCUMENTS_SEARCH_CLOSE_PIT',
+          };
+        }
+      }
+    } else {
+      throwBadResponse(stateP, res);
+    }
+  } else if (stateP.controlState === 'OUTDATED_DOCUMENTS_TRANSFORM') {
+    const res = resW as ExcludeRetryableEsError<ResponseType<typeof stateP.controlState>>;
+    if (Either.isRight(res)) {
+      // we haven't seen corrupt documents or any transformation errors thus far in the migration
+      // index the migrated docs
+      if (stateP.corruptDocumentIds.length === 0 && stateP.transformErrors.length === 0) {
         return {
           ...stateP,
-          controlState: 'OUTDATED_DOCUMENTS_SEARCH_CLOSE_PIT',
+          controlState: 'TRANSFORMED_DOCUMENTS_BULK_INDEX',
+          transformedDocs: [...res.right.processedDocs],
+          hasTransformedDocs: true,
+        };
+      } else {
+        // We have seen corrupt documents and/or transformation errors
+        // skip indexing and go straight to reading and transforming more docs
+        return {
+          ...stateP,
+          controlState: 'OUTDATED_DOCUMENTS_SEARCH_READ',
```

Comment:
Can we add an integration test to make sure Kibana reads **all** the SO (in several batches, preferably) and logs transformation errors for all of them?

Reply:
ts throws an error for declaring a `Symbol` as a string. Symbols are used elsewhere in the tests too.

Reply:
That would be a great test indeed! Can we change the batch size so that we can keep the test relatively small? I'd hate to have to create hundreds of thousands of saved objects!

Do you have any tricks on how to make life easier with 'hacking' SO's in the `zip` archives?

Reply:
> If that won't be enough

sound good

> do you have any other tricks on how to make life easier with 'hacking' SO's in the zip archives?

I added a note about creating an archive https://github.com/elastic/kibana/blob/33eecc297973e348b0fc4907f62170fe27d1cefa/packages/kbn-test/src/es/test_es_cluster.ts#L28-L36
I used the following script to fill the ES server with data. maybe we shoould clean it up and add to out `test_helpers`
```typescript
import { Client } from '@elastic/elasticsearch';
import Path from 'path';
import Fs from 'fs';
import Util from 'util';
import * as kbnTestServer from './src/core/test_helpers/kbn_server';
import type { SavedObjectsRawDoc } from './src/core/server';

const logFilePath = Path.join(__dirname, 'migration_test_kibana.log');

const asyncUnlink = Util.promisify(Fs.unlink);
async function removeLogFile() {
  // ignore errors if it doesn't exist
  await asyncUnlink(logFilePath).catch(() => void 0);
}

const client = new Client({
  node: 'http://elastic:changeme@localhost:9220',
});

async function startStack() {
  await removeLogFile();
  const { startES } = kbnTestServer.createTestServers({
    adjustTimeout: (t: number) => null,
    settings: {
      es: {
        license: 'trial',
      },
    },
  });
  const indexName = '.kibana_8.0.0_001';
  const versionAliasName = '.kibana_8.0.0';
  const aliasName = '.kibana';
  const esServer = await startES();
  await client.indices.create({
    index: indexName,
    // wait_for_active_shards: 'all',
    body: {
      mappings: {
        // dynamic: 'strict',
        properties: {
          migrationVersion: {
            dynamic: 'true',
            type: 'object',
          },
          type: {
            type: 'keyword',
          },
          namespace: {
            type: 'keyword',
          },
          namespaces: {
            type: 'keyword',
          },
          originId: {
            type: 'keyword',
          },
          updated_at: {
            type: 'date',
          },
          references: {
            type: 'nested',
            properties: {
              name: {
                type: 'keyword',
              },
              type: {
                type: 'keyword',
              },
              id: {
                type: 'keyword',
              },
            },
          },
          coreMigrationVersion: {
            type: 'keyword',
          },
        },
      },
      settings: {
        number_of_replicas: 0,
      },
    },
  });

  const { body } = await client.index({
    id: 'foo:my_name',
    index: indexName,
    refresh: 'wait_for',
    wait_for_active_shards: 'all',
    body: {
      type: 'foo',
      foo: {},
      migrationVersion: {
        foo: '7.13.0',
      },
    },
  });

  // setup alias
  await client.indices.putAlias({
    index: indexName,
    name: aliasName,
  });

  // setup alias
  await client.indices.putAlias({
    index: indexName,
    name: versionAliasName,
  });

  // make sure doc is indexed
  const res = await client.search({
    index: indexName,
    body: {
      sort: ['_doc'],
    },
    size: 10000,
  });
  const allDocuments = res.body.hits.hits as SavedObjectsRawDoc[];
  console.log('>>docs', allDocuments);

  await esServer.stop();
}
startStack();
```

Reply:
let's add an integration test in a follow-up

Reply:
I ran the script from `src/core/server/saved_objects/migrationsv2/integration_tests/` folder. To run it as a script you need an intermediate js file with:
```js
// intermediate js file
require('../src/setup_node_env');
require('../path_to_script');
```
and then in the console: `node path_to_intermediate.js`